swap/swap.go

```diff
@@ -1,98 +1,577 @@
+}
+
+// Add is the (sole) accounting function
+// Swap implements the protocols.Balance interface
 func (s *Swap) Add(amount int64, peer *protocols.Peer) (err error) {
-	s.lock.Lock()
-	defer s.lock.Unlock()
+	s.accountingLock.Lock()
+	defer s.accountingLock.Unlock()
 
-	//load existing balances from the state store
-	err = s.loadState(peer)
+	err = s.loadBalance(peer.ID())
 	if err != nil && err != state.ErrNotFound {
-		return
+		return fmt.Errorf("error while loading balance for peer %s", peer.ID().String())
+	}
+
+	// Check if balance with peer is over the disconnect threshold
+	balance, exists := s.getBalance(peer.ID())
+	if !exists {
+		return fmt.Errorf("peer %v does not exist", peer.ID())
 	}
+	if balance >= s.disconnectThreshold {
+		return fmt.Errorf("balance for peer %s is over the disconnect threshold %d, disconnecting", peer.ID().String(), s.disconnectThreshold)
+	}
+
+	var newBalance int64
+	newBalance, err = s.updateBalance(peer.ID(), amount)
+	if err != nil {
+		return err
+	}
+
+	// Check if balance with peer crosses the payment threshold
+	// It is the peer with a negative balance who sends a cheque, thus we check
+	// that the balance is *below* the threshold
+	if newBalance <= -s.paymentThreshold {
+		log.Warn("balance for peer went over the payment threshold, sending cheque", "peer", peer.ID().String(), "payment threshold", s.paymentThreshold)
+		swapPeer, ok := s.getPeer(peer.ID())
+		if !ok {
+			return fmt.Errorf("peer %s not found", peer)
+		}
+		return s.sendCheque(swapPeer)
+	}
+
+	return nil
+}
+
+func (s *Swap) getBalance(id enode.ID) (int64, bool) {
+	s.balancesLock.RLock()
+	defer s.balancesLock.RUnlock()
+	peerBalance, exists := s.balances[id]
+	return peerBalance, exists
+}
+
+func (s *Swap) setBalance(id enode.ID, balance int64) {
+	s.balancesLock.Lock()
+	defer s.balancesLock.Unlock()
+	s.balances[id] = balance
+}
+
+func (s *Swap) getCheque(id enode.ID) (*Cheque, bool) {
+	s.chequesLock.RLock()
+	defer s.chequesLock.RUnlock()
+	peerCheque, exists := s.cheques[id]
+	return peerCheque, exists
+}
+
+func (s *Swap) setCheque(id enode.ID, cheque *Cheque) {
+	s.chequesLock.Lock()
+	defer s.chequesLock.Unlock()
+	s.cheques[id] = cheque
+}
+
+// handleMsg is for handling messages when receiving messages
+func (s *Swap) handleMsg(p *Peer) func(ctx context.Context, msg interface{}) error {
+	return func(ctx context.Context, msg interface{}) error {
+		switch msg := msg.(type) {
+		case *EmitChequeMsg:
+			go s.handleEmitChequeMsg(ctx, p, msg)
+		}
+		return nil
+	}
+}
+
+// handleEmitChequeMsg should be handled by the creditor when it receives
+// a cheque from a debitor
+func (s *Swap) handleEmitChequeMsg(ctx context.Context, p *Peer, msg *EmitChequeMsg) error {
+	cheque := msg.Cheque
+	log.Info("received cheque from peer", "peer", p.ID().String())
+	actualAmount, err := s.processAndVerifyCheque(cheque, p)
+	if err != nil {
+		return err
```

Comment:
This error is never caught as this method is called in a goroutine.

I would suggest to decide which goroutines are needed and to handle errors at the topmost level, even if that means just logging them.